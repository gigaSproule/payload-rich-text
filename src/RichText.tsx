import {
  GraphQlRelationshipData,
  Relationship,
  RelationshipData,
} from "./Relationship";
import { GraphQlUploadData, Upload, UploadData } from "./Upload";
import type { BulletListData, CheckListData, NumberListData } from "./List";
import { List } from "./List";
import type { ParagraphData } from "./Paragraph";
import { Paragraph } from "./Paragraph";
import type { HeadingData } from "./Heading";
import { Heading } from "./Heading";
import type { QuoteData } from "./Quote";
import { Quote } from "./Quote";
import type { HorizontalRuleData } from "./HorizontalRule";
import { HorizontalRule } from "./HorizontalRule";
import type { Options } from "./types";
import { Block, BlockData, GraphQlBlockData } from "./Block";

export type RichTextRoot<T extends "strict" | "graphql" = "strict"> = {
  children: T extends "strict"
    ? (
        | ParagraphData
        | CheckListData
        | BulletListData
        | NumberListData
        | HeadingData
        | HorizontalRuleData
        | QuoteData
        | UploadData
        | RelationshipData
        | BlockData
      )[]
    : (
        | ParagraphData
        | CheckListData
        | BulletListData
        | NumberListData
        | HeadingData
        | HorizontalRuleData
        | QuoteData
        | GraphQlUploadData
        | GraphQlRelationshipData
        | GraphQlBlockData
      )[];
  direction: "ltr" | "rtl" | null;
  format: "";
  indent: 0;
  type: "root";
  version: 1;
};

export type RichTextData<T extends "strict" | "graphql" = "strict"> = {
  root: RichTextRoot<T>;
};

export type Props = {
  data: RichTextData | RichTextData<"graphql">;
  options?: Options;
};

export const RichText = ({ data, options }: Props) => {
  return (
    <>
      {data.root.children.flatMap((child, index) => {
        switch (child.type) {
          case "paragraph": {
            return <Paragraph key={index} data={child} options={options} />;
          }
          case "list": {
            return <List key={index} data={child} options={options} />;
          }
          case "heading": {
            return <Heading key={index} data={child} options={options} />;
          }
          case "horizontalrule": {
            return (
              <HorizontalRule key={index} data={child} options={options} />
            );
          }
          case "quote": {
            return <Quote key={index} data={child} options={options} />;
          }
          case "upload": {
            return <Upload key={index} data={child} options={options} />;
          }
          case "relationship": {
            return <Relationship key={index} data={child} options={options} />;
          }
          case "block": {
            return <Block key={index} data={child} options={options} />;
          }
          default: {
            return options?.unknown ? options.unknown(child) : undefined;
          }
        }
      })}
    </>
  );
};
