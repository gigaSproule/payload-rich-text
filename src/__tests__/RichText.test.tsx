import { afterAll, beforeAll, describe, expect, it, jest } from "@jest/globals";
import { render } from "@testing-library/react";
import type { Props } from "../RichText";
import { RichText } from "../RichText";
import { createParagraphData } from "./helpers/createParagraphData";
import {
  createRichTextData,
  createRichTextRoot,
} from "./helpers/createRichTextData";
import {
  createBulletListData,
  createCheckListData,
  createNumberListData,
} from "./helpers/createListData";
import { createQuoteData } from "./helpers/createQuoteData";
import { createRelationshipData } from "./helpers/createRelationshipData";
import { createBlockData } from "./helpers/createBlockData";
import { createHeadingData } from "./helpers/createHeadingData";
import { createHorizontalRuleData } from "./helpers/createHorizontalRuleData";
import { createUploadData } from "./helpers/createUploadData";

const originalCrypto = global.crypto;

beforeAll(() => {
  Object.defineProperty(global, "crypto", {
    value: {
      randomUUID: jest.fn().mockReturnValue("randomUUID"),
    },
    writable: true,
  });
});

afterAll(() => {
  Object.defineProperty(global, "crypto", {
    value: originalCrypto,
    writable: true,
  });
});

describe("RichText", () => {
  it("should not render anything if there are no children", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render paragraph if paragraph is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createParagraphData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render heading if heading is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createHeadingData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render horizontal rule if horizontal rule is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createHorizontalRuleData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render bullet list if bullet list is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createBulletListData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render check list if check list is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createCheckListData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render number list if number list is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createNumberListData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render quote if quote is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createQuoteData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render relationship if relationship is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createRelationshipData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render upload if upload is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createUploadData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should render block if block is a child", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        children: [createBlockData()],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should not render anything if unknown is a child and no options provided", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        // @ts-expect-error -- forced unknown type to handle edge case
        children: [{ type: "unknown" }],
      }),
    });
    const { container } = render(<RichText data={richTextData} />);
    expect(container).toMatchSnapshot();
  });

  it("should not render anything if unknown is a child and options provided without unknown", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        // @ts-expect-error -- forced unknown type to handle edge case
        children: [{ type: "unknown" }],
      }),
    });
    const { container } = render(<RichText data={richTextData} options={{}} />);
    expect(container).toMatchSnapshot();
  });

  it("should render custom render if unknown is a child and options provided with unknown", () => {
    const richTextData: Props["data"] = createRichTextData({
      root: createRichTextRoot({
        // @ts-expect-error -- forced unknown type to handle edge case
        children: [{ type: "unknown" }],
      }),
    });
    const { container } = render(
      <RichText
        data={richTextData}
        options={{
          unknown: (unknown) => {
            return (
              <div key={"unknown"}>
                Custom Unknown - {JSON.stringify(unknown)}
              </div>
            );
          },
        }}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
