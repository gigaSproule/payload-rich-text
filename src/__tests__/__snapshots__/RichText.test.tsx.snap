// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RichText should not render anything if there are no children 1`] = `<div />`;

exports[`RichText should not render anything if unknown is a child and no options provided 1`] = `<div />`;

exports[`RichText should not render anything if unknown is a child and options provided without unknown 1`] = `<div />`;

exports[`RichText should render block if block is a child 1`] = `<div />`;

exports[`RichText should render bullet list if bullet list is a child 1`] = `
<div>
  <ul
    class="list-bullet"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ul>
</div>
`;

exports[`RichText should render check list if check list is a child 1`] = `
<div>
  <ul
    class="list-check"
  >
    <li
      aria-checked="false"
      role="checkbox"
      tabindex="-1"
      value="1"
    >
      <input
        id="randomUUID"
        type="checkbox"
      />
      <label
        for="randomUUID"
      >
        heading text
      </label>
    </li>
  </ul>
</div>
`;

exports[`RichText should render custom render if unknown is a child and options provided with unknown 1`] = `
<div>
  <div>
    Custom Unknown - 
    {"type":"unknown"}
  </div>
</div>
`;

exports[`RichText should render heading if heading is a child 1`] = `
<div>
  <h1>
    heading text
  </h1>
</div>
`;

exports[`RichText should render horizontal rule if horizontal rule is a child 1`] = `
<div>
  <hr />
</div>
`;

exports[`RichText should render line break if line break is a child 1`] = `
<div>
  <br />
</div>
`;

exports[`RichText should render number list if number list is a child 1`] = `
<div>
  <ol
    class="list-number"
  >
    <li
      value="1"
    >
      heading text
    </li>
  </ol>
</div>
`;

exports[`RichText should render paragraph if paragraph is a child 1`] = `
<div>
  <p>
    <span>
      heading text
    </span>
  </p>
</div>
`;

exports[`RichText should render quote if quote is a child 1`] = `
<div>
  <blockquote>
    heading text
  </blockquote>
</div>
`;

exports[`RichText should render relationship if relationship is a child 1`] = `<div />`;

exports[`RichText should render upload if upload is a child 1`] = `
<div>
  <a
    href="https://www.example.com/image.jpg"
  >
    image.jpg
  </a>
</div>
`;
